### PR TITLE
Log top 5 largest files when TS language service is disabling.

### PR DESCRIPTION
**Pull request by unional**
This PR mitigates an issue on vscode: https://github.com/Microsoft/vscode/issues/35549

It will now log the top 5 largest files loaded.
Suggestion on how to format the message is welcome.

There is no test added 😞  because I can't find a way to test this private method, and also stubbing the logger to validate the output.

### CODE REVIEW

**Comment by unional** [comment]
If you find sorting and managing the top5 list is too time-consuming, maybe we can just report the largest file.

**Comment by unional** [comment]
I notice this method is being called twice on the same project. So it seems that it double counted the file size.

https://github.com/Microsoft/vscode/issues/35549#issuecomment-337761571

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -1372,19 +1372,37 @@ namespace ts.server {
             this.projectToSizeMap.forEach(val => (availableSpace -= (val || 0)));
 
             let totalNonTsFileSize = 0;
+            let top5LargestFiles = [] as { name: string, size: number }[];
```
Comment (DanielRosenwasser) [inline src/server/editorServices.ts]: Use an annotation, not a type assertion.

**Comment by DanielRosenwasser** [comment]
#19326 is probably the right issue to track this.

**Comment by DanielRosenwasser** [comment]
I think that in the error case, you should just sort the list of all files and slice off the top 5 instead of inserting the logic into the loop

**Comment by unional** [comment]
> I think that in the error case, you should just sort the list of all files and slice off the top 5 instead of inserting the logic into the loop.

Sure, can do that too. But in order to do that, I have to use `{ name: string, size: number}[n]` where `n` is number of files.
Just more memory to GC.

**Comment by unional** [comment]
Any comment on the log message?

**Comment by unional** [comment]
In the last commit, I use `logExceedLimit.call(this)` because I don't want to declare `const logExceedLimit = () => { ... }` in front of the method.

Error handling and logging code should not obstruct main code path.

File: src/server/editorServices.ts [hunk 2]
```diff
@@ -1372,24 +1372,50 @@ namespace ts.server {
...
+                const files: { name: string, size: number }[] = [];
+                let totalNonTsFileSize = 0;
+                for (const f of fileNames) {
+                    const name = propertyReader.getFileName(f);
+                    if (hasTypeScriptFileExtension(name)) {
+                        continue;
+                    }
+
+                    const size = this.host.getFileSize(name);
+                    totalNonTsFileSize += size;
```
Comment (mhegazy) [inline src/server/editorServices.ts]: do you need to compute this again?

Comment (mhegazy) [reply src/server/editorServices.ts]: consider:

```ts
fileNames
   .map(f => hasTypeScriptFileExtension(name) ? 0 : { name, size: host.getFileSize(name) })
   .sort((a, b) => b.size - a.size)
   .slice(0,5)
   .map(f => `${f.name} (${f.size})`)
   .join(",")
```

File: src/server/editorServices.ts [hunk 3]
```diff
@@ -1372,24 +1372,50 @@ namespace ts.server {
...
             for (const f of fileNames) {
                 const fileName = propertyReader.getFileName(f);
                 if (hasTypeScriptFileExtension(fileName)) {
                     continue;
                 }
+
                 totalNonTsFileSize += this.host.getFileSize(fileName);
+
                 if (totalNonTsFileSize > maxProgramSizeForNonTsFiles) {
+                    logExceedLimit.call(this);
```
Comment (mhegazy) [inline src/server/editorServices.ts]: do we need top 5? or can we jsut log the largest so far? if so, we can keep the largest file in a local.

Comment (unional) [reply src/server/editorServices.ts]: W can do that too. I initially do top 5 just to get more information so I can exclude/fix multiple files at the same time.

File: src/server/editorServices.ts (hunk 2 above)
Comment (unional) [reply src/server/editorServices.ts]: That's a good idea, but I'll separate formatting from the chain.

**Comment by unional** [comment]
> do you need to compute this again?

I change to passing that value in. However, it is not an accurate number because in the main loop it exits once `totalNonTsFileSize > maxProgramSizeForNonTsFiles`.

File: src/server/editorServices.ts [hunk 4]
```diff
@@ -1372,24 +1372,39 @@ namespace ts.server {
...
 
             if (totalNonTsFileSize > availableSpace) {
+                logExceedLimit.call(this, totalNonTsFileSize);
                 return true;
             }
 
             this.projectToSizeMap.set(name, totalNonTsFileSize);
             return false;
+
+            function logExceedLimit(this: ProjectService, totalNonTsFileSize: number) {
```
Comment (mhegazy) [inline src/server/editorServices.ts]: why do you need the `this` here?

Comment (mhegazy) [reply src/server/editorServices.ts]: just pass in the logger/host, or make the function return a string.

Comment (unional) [reply src/server/editorServices.ts]: Sure, that would be even better.

**Pull request merged**